internal/release/services/release/release.go

```diff
@@ -51,9 +68,174 @@ func (s *ReleaseService) RegisterWithServer(grpcServer grpc.ServiceRegistrar) er
+		return nil, MarshalReleaseError
+	}
+
+	in := &dynamodb.PutItemInput{
+		Item:      av,
+		TableName: aws.String(ReleaseTableName),
+	}
+
+	if _, err = s.Db.PutItem(ctx, in); err != nil {
+		span.RecordError(err)
+		log.Println(err)
+		return nil, PublishReleaseError
+	}
+
+	log.Println("New release created.")
+	return ReleasePublished, nil
+}
+
+// ListReleases Retrieves all releases.
+// Only releases that have been published should be reported.
+func (s *ReleaseService) ListReleases(ctx context.Context, request *releaseSvc.ListReleasesRequest) (*releaseSvc.ListReleasesResponse, error) {
+
+	if request.MaxAgeSeconds == nil {
+		request.MaxAgeSeconds = &DefaultMaxAgeSeconds
+	}
+
+	// Generate UTC string (Now - MaxAgeSeconds)
+	splitTimeISO := time.Unix(time.Now().UTC().Unix()-int64(*request.MaxAgeSeconds), 0).UTC().String()
+
+	// Construct the filter builder with a name and value.
+	filter := expression.Name("createdAt").GreaterThanEqual(expression.Value(splitTimeISO))
+	expr, err := expression.NewBuilder().WithFilter(filter).Build()
+	if err != nil {
+		log.Printf("Expression Builder failed creation: %v", err)
+		return nil, err
+	}
+
+	scanQueryInputs := &dynamodb.ScanInput{
+		ExpressionAttributeNames:  expr.Names(),
+		ExpressionAttributeValues: expr.Values(),
+		FilterExpression:          expr.Filter(),
+		TableName:                 aws.String(ReleaseTableName),
+	}
+
+	response, err := s.Db.Scan(ctx, scanQueryInputs)
+	if err != nil {
+		log.Printf("ScanInput failed: %v", err)
+		return nil, err
+	}
+	log.Println("Filtered Release Count: ", len(response.Items))
+
+	grpcResponse := releaseSvc.ListReleasesResponse{}
+	if response.Items != nil {
+		for _, item := range response.Items {
+
+			Release := &releaseSvc.Release{}
+			if err3 := attributevalue.UnmarshalMap(item, &Release); err3 != nil {
+				log.Printf("UnmarshalMap failed: %v", err3)
+				return nil, err3
+			}
+			grpcResponse.Releases = append(grpcResponse.Releases, Release)
+		}
+	}
+
+	// Sort list of releases based on createdAt field
+	grpcResponse.Releases = sortReleaseList(grpcResponse.Releases)
+
+	return &grpcResponse, nil
+}
+
+// Sort a slice of releaseSvc.Release struct, using sort.SliceStable method
+func sortReleaseList(releases []*releaseSvc.Release) []*releaseSvc.Release {
+	sort.SliceStable(releases, func(a, b int) bool {
+
+		TimeA, err := time.Parse(TimeFormatLayout, releases[a].CreatedAt)
+		if err != nil {
+			log.Printf("Failed to parse createdAt field: %v", err)
+		}
+
+		TimeB, err := time.Parse(TimeFormatLayout, releases[b].CreatedAt)
+		if err != nil {
+			log.Printf("Failed to parse createdAt field: %v", err)
+		}
+		return TimeA.Before(TimeB)
+	})
+
+	return releases
+}
+
+// TODO: OPTIMIZE!!!
+// GetLatestRelease retrieves the latest release.
+func (s *ReleaseService) GetLatestRelease(ctx context.Context, request *releaseSvc.ListReleasesRequest) (*releaseSvc.ListReleasesResponse, error) {
+	log.Printf("Getting latest published release.")
```

Comment:
The optimisation here is to add another table which only stores the most recent release for a unique tuple of (type,org,name) and is updated whenever a new release is published.
For simplicity we could store the same data as we store in the "Release" log table

Reply:
To be clear this doesn't need to be done in this story - we can add another US to do this optimisation later.

Reply:
Marking it as resolved. We'll add a ticket to the backlog.